dali/test/dali_operator_test.h

```diff
@@ -0,0 +1,180 @@
+      AddInputsToPipeline<Backend>(*pipeline_, "input");
+    }
+    AddOperatorToPipeline(*pipeline_, op_spec);
+    outputs_ = RunPipeline<Backend>(*pipeline_, op_spec);
+    ASSERT_TRUE(Verify(outputs_, anticipated_outputs));
+  }
+
+
+ private:
+
+  virtual std::vector<std::pair<Input, Shape>> SetInputs() const = 0;
+
+  virtual std::string SetOperator() const = 0; // TODO std::vector
+
+  virtual bool Verify(Output outputs, Output anticipated_outputs) const = 0;
+
+
+  void SetUp() final {
+    auto inp = SetInputs();
+    auto num_inputs = inp.size();
+    assert(num_inputs >= 0);
+    has_input_ = num_inputs != 0;
+    if (has_input_) {
+      inputs_ = inp[0].first;
+      input_shape_ = inp[0].second;
+    }
+    operator_name_ = SetOperator();
+    InitPipeline();
+  }
+
+
+  void TearDown() final {
+
+  }
+
+
+  void InitPipeline() {
+    if (!pipeline_.get()) {
+      pipeline_.reset(new Pipeline(batch_size_, num_threads_, 0));
+    }
+  }
+
+
+  template<typename Backend>
+  void AddInputsToPipeline(Pipeline &pipeline, std::string input_name) {
+    pipeline.AddExternalInput(input_name);
+
+    auto tensor_list = ToTensorList<Backend>(inputs_, input_shape_);
+
+    //debug
+    auto ptr = tensor_list.template tensor<float>(0);
+
+    pipeline.SetExternalInput(input_name, tensor_list);
+  }
+
+
+  void AddOperatorToPipeline(Pipeline &pipeline, OpSpec op_spec) {
+    pipeline.AddOperator(op_spec);
+  }
+
+
+//  template<typename Backend>
+  DeviceWorkspace CreateWorkspace() const {
+    DeviceWorkspace ws;
+    return ws;
+  }
+
+
+  /**
+   * TODO
+   * @tparam Backend
+   * @param spec
+   * @return
+   */
+  template<typename Backend>
+  Output RunPipeline(Pipeline &pipeline, OpSpec spec) {
+    vector<std::pair<string, string>> vecoutputs_;
+    for (int i = 0; i < spec.NumOutput(); ++i) {
+      vecoutputs_.emplace_back(spec.OutputName(i), spec.OutputDevice(i));
+    }
+    pipeline.Build(vecoutputs_);
+    pipeline.RunCPU();
+    pipeline.RunGPU();
+
+    auto workspace = CreateWorkspace();
+    pipeline.Outputs(&workspace);
+    return GetOutputsFromWorkspace<Backend>(workspace);
+  }
+
+
+  template<typename Backend>
+  Output GetOutputsFromWorkspace(DeviceWorkspace ws) {
+    //TODO
+//    auto i = ws.Output<Backend>(0);
+//    auto ptr = i->template data<Output>();
+//    return *ptr;
+  }
+
+
+  template<typename Backend>
+  OpSpec CreateOpSpec(std::string operator_name, Arguments operator_arguments) {
+    OpSpec opspec = OpSpec(operator_name);
+    for (const auto &arg : operator_arguments) {
+      opspec.AddArg(arg.first, arg.second);
+    }
+    if (has_input_) {
+      opspec.AddInput("input", detail::BackendStringName<Backend>());
+    }
+    opspec.AddOutput("output", detail::BackendStringName<Backend>());
+    return opspec;
+  }
+
+
+  Input inputs_;
+  Output outputs_;
+  Shape input_shape_;
+  std::string operator_name_;
+  std::unique_ptr<Pipeline> pipeline_;
+  const size_t batch_size_ = 32;
```

Comment:
By now batch size and num threads can remain as they are, however it would be interesting to think about a customization point for them

Reply:
I left num_threads as they are, but changed batch_size policy. For now, class assumes, that entire input will be a single batch. It's not perfect, but enables easy modification in future. I just want to have it merged, so it can be used and further worked on